### PR TITLE
SONARPY-1251 S6542 should not raise an issue on overrides and overloads.

### PR DESCRIPTION
**Pull request by jdodinh**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java [hunk 1]
```diff
@@ -42,14 +52,52 @@ public void initialize(Context context) {
 
   private static void checkForAnyInTypeHint(SubscriptionContext ctx) {
     TypeAnnotation typeAnnotation = (TypeAnnotation) ctx.syntaxNode();
-    if (isTypeAny(typeAnnotation)) {
-      ctx.addIssue(typeAnnotation.expression(), MESSAGE);
-    }
+    Optional.of(typeAnnotation)
+      .filter(UseOfAnyAsTypeHintCheck::isTypeAny)
+      .filter(Predicate.not(UseOfAnyAsTypeHintCheck::isAnnotatingArgumentToChildMethod))
+      .filter(Predicate.not(UseOfAnyAsTypeHintCheck::isUnannotatedOverride))
```
Comment (jdodinh) [inline python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java:58]: Feel free to remove the negation of the predicate here. For some reason I just really didn't want to have `isNot...` at the beginning of the method name.

File: its/ruling/src/test/resources/expected/python-S6542.json [hunk 2]
```diff
@@ -103,68 +111,69 @@
...
-'project:mypy-0.782/test-data/stdlib-samples/3.2/pprint.py':[
+"project:mypy-0.782/test-data/stdlib-samples/3.2/pprint.py": [
 89,
 92,
-92,
+92
 ],
-'project:mypy-0.782/test-data/stdlib-samples/3.2/random.py':[
+"project:mypy-0.782/test-data/stdlib-samples/3.2/random.py": [
+100,
```
Comment (joke1196) [inline its/ruling/src/test/resources/expected/python-S6542.json:141]: This is a new issue detected

File: its/ruling/src/test/resources/expected/python-S6542.json [hunk 3]
```diff
@@ -1,64 +1,72 @@
...
 202,
 205,
 208,
 211,
-253,
+253
+],
+"project:mypy-0.782/mypy/main.py": [
+266,
+266
```
Comment (joke1196) [inline its/ruling/src/test/resources/expected/python-S6542.json:42]: These two are also new issues.

File: python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java [hunk 4]
```diff
@@ -32,24 +45,81 @@
...
   }
+
+  private static boolean hasFunctionOverrideOrOverloadDecorator(FunctionDef currentFunctionDef) {
+    return currentFunctionDef.decorators().stream()
+      .map(Decorator::expression)
+      .map(TreeUtils::fullyQualifiedNameFromExpression)
+      .anyMatch(OVERRIDE_FQNS::contains);
+  }
+
+  private static boolean isParentFunctionAnOverrideWithoutDecorators(FunctionDef currentMethodDef) {
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java:86]: Could we achieve the same using the `FunctionSymbolImpl#owner` method? Though I know it's annoying as it's not part of the public API.

Also, maybe this is too lenient, but we have the helper `SymbolUtils#canBeAnOverridingMethod` which could also be used here. I assume we'd lose many issues, as it is much more lenient, but that might prevent quite a few FPs when overriding methods from other libraries.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Looks good overall.

I think we shouldn't be so strict when checking for a match between the child and parents' parameters.

I also left a suggestion that would make the rule more lenient (assuming an override whenever the class hierarchy is unresolved), happy to discuss this.

File: python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java (hunk 4 above)
Comment (joke1196) [reply python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java:86]: Yes this is great. Thank you!

File: python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java [hunk 5]
```diff
@@ -20,36 +20,65 @@
...
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java:78]: Probably now the name `isParentFunctionAnOverrideWithoutDecorators` doesn't make so much sense anymore and could be simplified to `canParentFunctionBeAnOverride` or something like that.

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
Looks good! Minor renaming suggestion.

**Pull request merged**